### PR TITLE
Ea-49 Feature: contract manager

### PR DESCRIPTION
**Pull request by EnQs**

### CODE REVIEW

File: src/main/java/enqs/contract/Contract.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                getWorkNormType() == that.getWorkNormType() &&
+                getSalaryType() == that.getSalaryType() &&
+                Objects.equals(getWorkingDays(), that.getWorkingDays()) &&
+                Objects.equals(getHolidays(), that.getHolidays());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getWorkNormHours(), getWorkNormType(), getSalary(), getSalaryType(), getBillingDay(), getWorkingDays(), getHolidays());
+    }
```
Comment (EnQs) [inline src/main/java/enqs/contract/Contract.java]: YAGNI

File: src/main/java/enqs/contract/ContractService.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        this.contractRepository = contractRepository;
+    }
+
+    public void createContract(Contract contract) {
+        contractRepository.create(contract);
+    }
+
+    public Contract getContract() {
+        return contractRepository.get();
+    }
```
Comment (EnQs) [inline src/main/java/enqs/contract/ContractService.java]: Make repo exception when no contract is saved

File: src/main/java/enqs/contract/ContractService.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    private final ContractRepository contractRepository;
+
+    @SuppressWarnings("unused")
+    public ContractService(ContractRepository contractRepository) {
+        this.contractRepository = contractRepository;
+    }
+
+    public void createContract(Contract contract) {
+        contractRepository.create(contract);
+    }
```
Comment (EnQs) [inline src/main/java/enqs/contract/ContractService.java]: Make dedicated exception when contract already exists

File: src/main/java/enqs/contract/ContractService.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        return contractRepository.get();
+    }
+
+    public void editContract(Contract contract) {
+        contractRepository.edit(contract);
+    }
+
+    public void deleteContract() {
+        contractRepository.delete();
+    }
```
Comment (EnQs) [inline src/main/java/enqs/contract/ContractService.java]: Make dedicated exception when there is no file to delete

File: src/main/java/enqs/contract/repos/TextFileContractRepo.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        if (!FILE.exists()) {
+            throw new ContractRepositoryException(String
+                    .format("Can't %s fle, it does not exist.", getCallerMethodName()));
+        }
+    }
+
+    private String getCallerMethodName() {
+        return Thread.currentThread().getStackTrace()[3].getMethodName();
+    }
+}
```
Comment (EnQs) [inline src/main/java/enqs/contract/repos/TextFileContractRepo.java]: Move those exceptions to service

File: pom.xml [hunk 6]
```diff
@@ -16,17 +16,26 @@
...
             <artifactId>spring-boot-starter-web</artifactId>
             <version>2.2.2.RELEASE</version>
         </dependency>
 
+        <!-- https://mvnrepository.com/artifact/javax.json.bind/javax.json.bind-api -->
+        <dependency>
+            <groupId>javax.json.bind</groupId>
+            <artifactId>javax.json.bind-api</artifactId>
+            <version>1.0</version>
+        </dependency>
```
Comment (Tyvrel) [inline pom.xml]: use jackson json

File: pom.xml [hunk 7]
```diff
@@ -16,17 +16,26 @@
...
+
+        <!-- https://mvnrepository.com/artifact/org.eclipse/yasson -->
+        <dependency>
+            <groupId>org.eclipse</groupId>
+            <artifactId>yasson</artifactId>
+            <version>1.0.6</version>
+        </dependency>
```
Comment (Tyvrel) [inline pom.xml]: what is that?

File: src/main/java/enqs/contract/Contract.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+@SuppressWarnings("unused")
+public class Contract {
+    private Double workNormHours;
+
+    private WorkNormType workNormType;
+    private double salary;
+    private SalaryType salaryType;
+    private int billingDay;
+    private List<DayOfWeek> workingDays;
+    private List<LocalDate> holidays;
```
Comment (Tyvrel) [inline src/main/java/enqs/contract/Contract.java]: holidays sounds like dayOffs. Maybe some better name?

File: src/main/java/enqs/contract/Contract.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+import java.util.List;
+
+@SuppressWarnings("unused")
+public class Contract {
+    private Double workNormHours;
+
+    private WorkNormType workNormType;
+    private double salary;
+    private SalaryType salaryType;
+    private int billingDay;
```
Comment (Tyvrel) [inline src/main/java/enqs/contract/Contract.java]: billingDayOfMonth

File: src/main/java/enqs/contract/ContractService.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+    private void validateContractExist() {
+        if (contractRepository.get() == null) {
+            throw new NotFoundException(String
+                    .format("Can't %s, it does not exist.", getCallerMethodName()));
+        }
+    }
+
+    private String getCallerMethodName() {
+        return Thread.currentThread().getStackTrace()[3].getMethodName();
```
Comment (Tyvrel) [inline src/main/java/enqs/contract/ContractService.java]: `[3]` is an optimisic assumtion

File: src/main/java/enqs/contract/repos/TextFileContractRepo.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Override
+    public void create(Contract contract) {
+        serializeToJsonFile(contract);
+    }
+
+    @Override
+    public Contract get() {
+        return FILE.exists() ? deserializeFromJsonFile() : null;
```
Comment (Tyvrel) [inline src/main/java/enqs/contract/repos/TextFileContractRepo.java]: never return null!

File: src/main/java/enqs/worklog/repos/WorkLogRepositoryException.java [hunk 12]
```diff
@@ -0,0 +1,12 @@
+package enqs.worklog.repos;
+
+public class WorkLogRepositoryException extends RuntimeException {
```
Comment (Tyvrel) [inline src/main/java/enqs/worklog/repos/WorkLogRepositoryException.java]: you can combine `WorkLogRepositoryException` and `ContractRepositoryException` into one `RepositoryException`

**Pull request merged**